### PR TITLE
Loop cleaning

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
This is quite a big PR.

it tries to give a more strieght forward unified approach to how to do things in `bake_step_*()` functions.

The gist of this, is that for many cases, we want to have the functions use the following structure

```r
for (col_name in col_names) {
  do something with  new_data[[col_name]]
}
```

It also has the benefit that the column names `col_names` are only extracted once, so we will gain a little speed and consistency that way.

Getting rid of `map()` inside `bake_step_*()` functions is also beneficial since they mess up the error handling/passing we are trying to use.

# Todo

The following steps needs these changes but for various reason didn't get done now. I'm planning on addressing these in separate PRs first.

- [x] `step_window()` https://github.com/tidymodels/recipes/pull/1172
    - is a 1 to many step, but doens’t act like it
- [x] `step_regex()` https://github.com/tidymodels/recipes/pull/1169
    - similar to step_dummy(), missing keep_orginal_cols()
- [x] `step_ratio()` https://github.com/tidymodels/recipes/pull/1173
    - doesn’t feel right
- [x] `step_profile()` https://github.com/tidymodels/recipes/pull/1177
    - also feels weird
- [x] `step_ns()` https://github.com/tidymodels/recipes/pull/1178
    - more work
- [x] `step_lag()` https://github.com/tidymodels/recipes/pull/1179
    - is a 1 to many step, but doesn’t act like it
- [x] `step_impute_linear()`, `step_impute_knn()`, `step_impute_bag()` https://github.com/tidymodels/recipes/pull/1184
    - missing lots of things. such as recipes_eval_select()
- [x] `step_depth()` https://github.com/tidymodels/recipes/pull/1180
    - missing check_new_data()
- [x] `step_count() ` https://github.com/tidymodels/recipes/pull/1169
    - is similar to step_regex()
- [x] `step_classdist()` https://github.com/tidymodels/recipes/pull/1183
    - similar to step_dummy() missing keep_orginal_cols()
- [x] `step_bs()` https://github.com/tidymodels/recipes/pull/1178
    - more work
- [x] `step_dummy()` https://github.com/tidymodels/recipes/pull/1182

## example benchmark

```r
old <- function(new_data, object) {
  if (length(object$columns) > 0) {
    for (i in seq_along(object$columns)) {
      new_data[, object$columns[i]] <-
        sample(getElement(new_data, object$columns[i]))
    }
  }
  new_data
}

new <- function(new_data, object) {
  col_names <- object$columns
  for (col_name in col_names) {
    new_data[[col_name]] <- sample(new_data[[col_name]])
  }
  new_data
}

object <- list(columns = colnames(mtcars))

bench::mark(
  check = FALSE,
  old(mtcars, object),
  new(mtcars, object)
)
#> # A tibble: 2 × 6
#>   expression               min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>          <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 old(mtcars, object)  158.5µs    167µs     5772.    4.34MB     41.4
#> 2 new(mtcars, object)   97.2µs    104µs     9231.  169.21KB     41.9
```

### CODE REVIEW

File: R/isomap.R [hunk 1]
```diff
@@ -192,21 +192,22 @@ prep.step_isomap <- function(x, training, info = NULL, ...) {
 
 #' @export
 bake.step_isomap <- function(object, new_data, ...) {
-  check_new_data(names(object$columns), object, new_data)
+  col_names <- names(object$columns)
+  check_new_data(col_names, object, new_data)
 
-  if (object$num_terms > 0 && length(object$columns) > 0L) {
-    isomap_vars <- colnames(environment(object$res@apply)$indata)
```
Comment (simonpcouch) [inline R/isomap.R:198]: Poked around a bit in blames and saw that `isomap_vars` used to be used everywhere where `names(object$columns)` now is, so the distinction between the two indeed is not needed. :)

File: R/kpca.R [hunk 2]
```diff
@@ -162,15 +163,15 @@ bake.step_kpca <- function(object, new_data, ...) {
...
       object = object$res,
-      rlang::expr(as.matrix(new_data[, object$columns]))
+      rlang::expr(as.matrix(new_data[, col_names]))
     )
   comps <- rlang::eval_tidy(cl)
   comps <- comps[, seq_len(object$num_comp), drop = FALSE]
   colnames(comps) <- names0(ncol(comps), object$prefix)
   comps <- as_tibble(comps)
   comps <- check_name(comps, new_data, object)
   new_data <- vec_cbind(new_data, comps)
```
Comment (simonpcouch) [inline R/kpca.R:173]:
```suggestion
  new_data <- vctrs::vec_cbind(new_data, comps)
```

File: R/kpca_poly.R [hunk 3]
```diff
@@ -169,15 +170,15 @@ bake.step_kpca_poly <- function(object, new_data, ...) {
...
       object = object$res,
-      rlang::expr(as.matrix(new_data[, object$columns]))
+      rlang::expr(as.matrix(new_data[, col_names]))
     )
   comps <- rlang::eval_tidy(cl)
   comps <- comps[, seq_len(object$num_comp), drop = FALSE]
   colnames(comps) <- names0(ncol(comps), object$prefix)
   comps <- as_tibble(comps)
   comps <- check_name(comps, new_data, object)
   new_data <- vec_cbind(new_data, comps)
```
Comment (simonpcouch) [inline R/kpca_poly.R:180]:
```suggestion
  new_data <- vctrs::vec_cbind(new_data, comps)
```

We may be better off doing a global search and replace on this, so I'll stop leaving this comment. :)

File: R/log.R [hunk 4]
```diff
@@ -118,32 +118,34 @@ prep.step_log <- function(x, training, info = NULL, ...) {
...
-    for (i in seq_along(col_names)) {
-      new_data[[col_names[i]]] <-
-        ifelse(abs(new_data[[col_names[i]]]) < 1,
-          0,
-          sign(new_data[[col_names[i]]]) *
-            log(abs(new_data[[col_names[i]]]), base = object$base)
-        )
+  if (object$signed && object$offset != 0) {
+    rlang::warn("When signed is TRUE, offset will be ignored")
+  }
```
Comment (simonpcouch) [inline R/log.R:131]: I like this refactor!!!

File: R/normalize.R [hunk 5]
```diff
@@ -151,13 +151,16 @@ prep.step_normalize <- function(x, training, info = NULL, ...) {
...
 bake.step_normalize <- function(object, new_data, ...) {
-  check_new_data(names(object$means), object, new_data)
+  col_names <- names(object$means)
+  check_new_data(col_names, object, new_data)
 
-  for (column in names(object$means)) {
-    mean <- object$means[column]
-    sd <- object$sds[column]
-    new_data[[column]] <- (new_data[[column]] - mean) / sd
+  for (col_names in col_names) {
```
Comment (simonpcouch) [inline R/normalize.R:157]: Could we `col_name in col_names` here? The two uses of `col_names` have me 🤯

File: R/ordinalscore.R [hunk 6]
```diff
@@ -118,14 +118,15 @@ prep.step_ordinalscore <- function(x, training, info = NULL, ...) {
...
+  check_new_data(col_names, object, new_data)
 
-  scores <- lapply(new_data[, object$columns], object$convert)
-  scores <- lapply(scores, vec_cast, integer())
-
-  for (i in object$columns) {
-    new_data[[i]] <- scores[[i]]
+  for (col_name in col_names) {
+    score <- object$convert(new_data[[col_name]])
+    score <- vec_cast(score, integer())
```
Comment (simonpcouch) [inline R/ordinalscore.R:126]:
```suggestion
    score <- vctrs::vec_cast(score, integer())
```

File: R/relu.R [hunk 7]
```diff
@@ -141,13 +141,15 @@ prep.step_relu <- function(x, training, info = NULL, ...) {
...
+  check_new_data(col_names, object, new_data)
 
   make_relu_call <- function(col) {
     call2("relu", sym(col), object$shift, object$reverse, object$smooth)
   }
-  exprs <- purrr::map(object$columns, make_relu_call)
-  newname <- glue("{object$prefix}{object$columns}")
+
+  exprs <- purrr::map(col_names, make_relu_call)
+  newname <- glue("{object$prefix}{col_names}")
```
Comment (simonpcouch) [inline R/relu.R:152]:
```suggestion
  newname <- glue::glue("{object$prefix}{col_names}")
```

File: R/time.R [hunk 8]
```diff
@@ -130,20 +130,21 @@ prep.step_time <- function(x, training, info = NULL, ...) {
...
-  for (column in object$columns) {
+  for (col_name in col_names) {
     time_values <- get_time_features(
-      dt = new_data[[column]],
+      dt = new_data[[col_name]],
       feats = object$features
     )
 
-    names(time_values) <- glue("{column}_{names(time_values)}")
+    names(time_values) <- glue("{col_name}_{names(time_values)}")
```
Comment (simonpcouch) [inline R/time.R:142]:
```suggestion
    names(time_values) <- glue::glue("{col_name}_{names(time_values)}")
```

**Review by simonpcouch — COMMENTED** [review]
Looking good! Not approving merge quite yet as I am fatigued and will work through these again in reverse order after you've addressed comments🤪

File: R/class.R [hunk 9]
```diff
@@ -184,14 +184,12 @@ bake_check_class_core <- function(x,
...
-bake.check_class <- function(object,
-                             new_data,
-                             ...) {
+bake.check_class <- function(object, new_data, ...) {
   col_names <- names(object$class_list)
-
   check_new_data(col_names, object, new_data)
 
-  mapply(bake_check_class_core,
+  mapply(
```
[the remark this answers is not in the document]
Comment (EmilHvitfeldt) [reply R/class.R:191]: This is intended

File: R/invlogit.R [hunk 10]
```diff
@@ -85,17 +85,13 @@ prep.step_invlogit <- function(x, training, info = NULL, ...) {
...
 #' @export
 bake.step_invlogit <- function(object, new_data, ...) {
-  check_new_data(names(object$columns), object, new_data)
+  col_names <- names(object$columns)
+  check_new_data(col_names, object, new_data)
 
-  for (i in seq_along(object$columns)) {
-    new_data[[object$columns[i]]] <-
-      binomial()$linkinv(
-        unlist(
```
[the remark this answers is not in the document]
Comment (EmilHvitfeldt) [reply R/invlogit.R:93]: I think it is a weird leftover thing. `new_data` is a data.frame. calling `new_data[[col_name]]` produces a column. Calling `unlist()` on said column doesn't do anything

File: R/log.R (hunk 4 above)
Comment (EmilHvitfeldt) [reply R/log.R:131]: Yes this is a great one!

File: R/normalize.R (hunk 5 above)
Comment (EmilHvitfeldt) [reply R/normalize.R:157]: It technically works, but I'm gonna fix it anyways :D 

``` r
x <- 1:5

for (x in x) {
  cat(x, "\n")
}
#> 1 
#> 2 
#> 3 
#> 4 
#> 5

x
#> [1] 5
```

<sup>Created on 2023-08-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Review by simonpcouch — APPROVED** [review]
Okay, I'm on board! :)

So many of these really do feel so much cleaner after this rewrite. Bravo!

**Pull request merged**